File: matlab/store.m
function store(ctx)
    if ctx.i > 1;
        % update J, u and y
        logfile = matfile(ctx.logfile, 'writable', true);
        logfile.J(:,ctx.i) = ctx.J;
        logfile.u(:,ctx.i) = ctx.u';        % transposed
        logfile.y(:,:,ctx.i) = ctx.y;
    else;
        % we are at the beginning of the optimization and
        % store initially J, lb, u, ub, ud, y, yd
        J   = ctx.J;
        lb  = ctx.lb';                      % transposed
        u   = ctx.u';                       % transposed
        ub  = ctx.ub';                      % transposed
        ud  = ctx.ud';                      % transposed
        % this is a trick, we make the stored variable 3D,
        % so it can be extended during the other iteratons
        y   = cat(3,ctx.y,ctx.y);
        yd  = ctx.yd;
        save(ctx.logfile, 'J', 'lb', 'u', 'ub', 'ud', 'y', 'yd', '-v7.3');
    end;
end

File: python/store.py
import h5py

def store(ctx):
    if ctx.i==1:
        # we are at the beginning of the optimization and
        # store initially J, lb, u, ub, ud, y, yd
        logfile = h5py.File(ctx.logfile, "w")
        logfile.create_dataset("J", data=ctx.J[None,:], maxshape=(None,ctx.ndata))
        logfile.create_dataset("lb", data=ctx.lb[None,:])
        logfile.create_dataset("u", data=ctx.u[None,:], maxshape=(None,ctx.nu))
        logfile.create_dataset("ub", data=ctx.ub[None,:])
        logfile.create_dataset("ud", data=ctx.ud[None,:])
        logfile.create_dataset("y", data=ctx.y[None,:,:], maxshape=(None,ctx.ndata,ctx.nx), compression="gzip")
        logfile.create_dataset("yd", data=ctx.yd, compression="gzip")
        logfile.close()
    else:
        # update J, u and y
        logfile = h5py.File(ctx.logfile, "a")
        logfile["J"].resize(ctx.i, axis=0)
        logfile["J"][ctx.i-1,:] = data=ctx.J
        logfile["u"].resize(ctx.i, axis=0)
        logfile["u"][ctx.i-1,:] = data=ctx.u
        logfile["y"].resize(ctx.i, axis=0)
        logfile["y"][ctx.i-1,:,:] = data=ctx.y
        logfile.close()


